server/src/main/java/io/druid/client/DruidDataSource.java

```diff
@@ -38,8 +38,8 @@
 
   private final String name;
   private final Map<String, String> properties;
-  private final Map<String, DataSegment> partitionNames;
-  private final ConcurrentSkipListSet<DataSegment> segmentsHolder;
+  private final Map<String, DataSegment> idToSegmentMap;
+  private final Set<DataSegment> segmentsHolder;
```

Comment:
At a cursory glance it doesn't even look like we need segmentsHolder. A `ConcurrentHashMap` for `idToSegmentMap` should suffice for all use cases. This would also remove the "Big Kernel Lock" in this class.

Reply:
I took an opportunity to give a sho at https://github.com/druid-io/druid/pull/5056 , which changes the concurrency pattern in DruidServer. I think it could be changed here as well.

Reply:
Good point. I removed segmentsHolder and changed idToSegmentMap to ConcurrentHashMap. Also, thanks for raising #5056.